paddle/fluid/framework/reader.h

```diff
@@ -31,9 +32,24 @@ class ReaderBase {
   virtual void ReInit() = 0;
 
   virtual ~ReaderBase();
+
+  // Return the readers which are the end of decorating chain. Basically
+  // they are readers just before read op.
+  std::unordered_set<ReaderBase*> GetEndPoints();
+
+ private:
+  friend class DecoratedReader;
+  // These methods can be only invoked inside DecoratedReader to record the
+  // decorating chain.
+  void InsertDecoratedReader(
+      const std::shared_ptr<ReaderBase>& decorated_reader);
+  // A set of which readers that decorated this reader.
+  std::vector<std::weak_ptr<ReaderBase>> decorated_readers_;
```

Comment:
I wonder, is it better to change the type of `decorated_readers_ ` to be `std::vector<std::weak_ptr<DecoratedReader>>` to avoid some possibly type mismatch error.

Reply:
OK, cool

Reply:
When declaring `class ReaderBase`, the `DecoratedReader` has not been declared.  We cannot use std::shared_ptr<T> or std::weak_ptr<T> if they are not declared.

The InsertDecoratedReader is a private method, only its friend class, `DecoratedReader` can invoke this method.